File: mobile/lib/models/signal.dart
class Signal {
  final int id;
  final String ticker;
  final String stockName;
  final String signalType;
  final double scoreTotal;
  final double? currentPrice;
  final double? entryLow;
  final double? entryHigh;
  final double? stopLoss;
  final double? targetLow;
  final double? targetHigh;
  final String? riskLevel;
  final int? horizonDays;
  final SignalBreakdown? breakdown;
  final DateTime createdAt;

  Signal({
    required this.id,
    required this.ticker,
    required this.stockName,
    required this.signalType,
    required this.scoreTotal,
    this.currentPrice,
    this.entryLow,
    this.entryHigh,
    this.stopLoss,
    this.targetLow,
    this.targetHigh,
    this.riskLevel,
    this.horizonDays,
    this.breakdown,
    required this.createdAt,
  });

  factory Signal.fromJson(Map<String, dynamic> json) => Signal(
        id: json['id'] is int ? json['id'] : int.parse(json['id'].toString()),
        ticker: json['ticker'] ?? '',
        stockName: json['stockName'] ?? '',
        signalType: json['signalType'] ?? 'WATCH',
        scoreTotal: (json['scoreTotal'] as num?)?.toDouble() ?? 0,
        currentPrice: (json['currentPrice'] as num?)?.toDouble(),
        entryLow: (json['entryLow'] as num?)?.toDouble(),
        entryHigh: (json['entryHigh'] as num?)?.toDouble(),
        stopLoss: (json['stopLoss'] as num?)?.toDouble(),
        targetLow: (json['targetLow'] as num?)?.toDouble(),
        targetHigh: (json['targetHigh'] as num?)?.toDouble(),
        riskLevel: json['riskLevel'] as String?,
        horizonDays: json['horizonDays'] as int?,
        breakdown: json['breakdown'] != null
            ? SignalBreakdown.fromJson(json['breakdown'])
            : null,
        createdAt: DateTime.tryParse(json['createdAt'] ?? '') ?? DateTime.now(),
      );

  String get signalTypeDisplay => switch (signalType) {
        'BUY_TODAY' => 'BUY TODAY',
        'WATCH' => 'WATCH',
        'AVOID' => 'AVOID',
        'HIGH_RISK' => 'HIGH RISK',
        _ => signalType.replaceAll('_', ' '),
      };

  bool get isBuy => signalType == 'BUY_TODAY';

  String? get entryRange {
    if (entryLow == null || entryHigh == null) return null;
    return '\$${entryLow!.toStringAsFixed(0)}–\$${entryHigh!.toStringAsFixed(0)}';
  }

  String? get targetRange {
    if (targetLow == null || targetHigh == null) return null;
    return '\$${targetLow!.toStringAsFixed(0)}–\$${targetHigh!.toStringAsFixed(0)}';
  }

  String? get stopDisplay =>
      stopLoss != null ? '\$${stopLoss!.toStringAsFixed(0)}' : null;

  String? get priceDisplay =>
      currentPrice != null ? '\$${currentPrice!.toStringAsFixed(2)}' : null;
}

class SignalBreakdown {
  final double momentum;
  final double volume;
  final double catalyst;
  final double fundamental;
  final double sentiment;
  final double trend;
  final double risk;
  final String? explanationJson;
  final String? whyNowSummary;

  SignalBreakdown({
    required this.momentum,
    required this.volume,
    required this.catalyst,
    required this.fundamental,
    required this.sentiment,
    required this.trend,
    required this.risk,
    this.explanationJson,
    this.whyNowSummary,
  });

  factory SignalBreakdown.fromJson(Map<String, dynamic> json) =>
      SignalBreakdown(
        momentum: (json['momentumScore'] as num?)?.toDouble() ?? 50,
        volume: (json['relVolumeScore'] as num?)?.toDouble() ?? 50,
        catalyst: (json['newsScore'] as num?)?.toDouble() ?? 50,
        fundamental: (json['fundamentalsScore'] as num?)?.toDouble() ?? 50,
        sentiment: (json['sentimentScore'] as num?)?.toDouble() ?? 50,
        trend: (json['trendScore'] as num?)?.toDouble() ?? 50,
        risk: (json['riskScore'] as num?)?.toDouble() ?? 50,
        explanationJson: json['explanationJson'] as String?,
        whyNowSummary: json['whyNowSummary'] as String?,
      );
}

File: backend/Fintrest.Api/Models/Signal.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace Fintrest.Api.Models;

public enum SignalType
{
    [System.Runtime.Serialization.EnumMember(Value = "BUY_TODAY")]
    BUY_TODAY,
    [System.Runtime.Serialization.EnumMember(Value = "WATCH")]
    WATCH,
    [System.Runtime.Serialization.EnumMember(Value = "AVOID")]
    AVOID,
    [System.Runtime.Serialization.EnumMember(Value = "TAKE_PROFIT")]
    TAKE_PROFIT,
    [System.Runtime.Serialization.EnumMember(Value = "HIGH_RISK")]
    HIGH_RISK
}

[Table("scan_runs")]
public class ScanRun
{
    [Key]
    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
    public long Id { get; set; }

    [Required, MaxLength(50)]
    public string RunType { get; set; } = "daily";

    [MaxLength(20)]
    public string? MarketSession { get; set; }

    [MaxLength(20)]
    public string StrategyVersion { get; set; } = "v1.0";

    public DateTime StartedAt { get; set; } = DateTime.UtcNow;
    public DateTime? CompletedAt { get; set; }
    public int? UniverseSize { get; set; }
    public int SignalsGenerated { get; set; }

    [MaxLength(20)]
    public string Status { get; set; } = "running"; // running, success, failed

    public ICollection<Signal> Signals { get; set; } = [];
}

[Table("signals")]
public class Signal
{
    [Key]
    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
    public long Id { get; set; }

    public long StockId { get; set; }
    public long ScanRunId { get; set; }

    public double ScoreTotal { get; set; }

    [Required]
    public SignalType SignalType { get; set; }

    [MaxLength(20)]
    public string? StrategyVersion { get; set; }

    // Trade zone
    public double? EntryLow { get; set; }
    public double? EntryHigh { get; set; }
    public double? StopLoss { get; set; }
    public double? TargetLow { get; set; }
    public double? TargetHigh { get; set; }

    [MaxLength(20)]
    public string? RiskLevel { get; set; }

    public int? HorizonDays { get; set; }

    [MaxLength(20)]
    public string? Status { get; set; }

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime? ClosedAt { get; set; }

    // Navigation
    [ForeignKey(nameof(StockId))]
    public Stock Stock { get; set; } = null!;

    [ForeignKey(nameof(ScanRunId))]
    public ScanRun ScanRun { get; set; } = null!;

    public SignalBreakdown? Breakdown { get; set; }
    public ICollection<SignalEvent> Events { get; set; } = [];
}

[Table("signal_breakdowns")]
public class SignalBreakdown
{
    [Key]
    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
    public long Id { get; set; }

    public long SignalId { get; set; }

    // 8 factor scores (0-100)
    public double MomentumScore { get; set; }
    public double RelVolumeScore { get; set; }
    public double NewsScore { get; set; }
    public double FundamentalsScore { get; set; }
    public double SentimentScore { get; set; }
    public double TrendScore { get; set; }
    public double RiskScore { get; set; }
    // Smart Money family — 25% of composite (matches TipRanks Smart Score
    // ceiling). Composed of Insider 35% / Institutional 25% / Short 15%
    // / Congressional 15% / Options 10%. Defaults to 50 (neutral) when
    // sub-signals are missing so tickers without smart-money data don't
    // crater. Persisted per signal for explainability.
    public double SmartMoneyScore { get; set; } = 50.0;

    // JSON explanation
    [Column(TypeName = "jsonb")]
    public string? ExplanationJson { get; set; }

    public string? WhyNowSummary { get; set; }

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    [ForeignKey(nameof(SignalId))]
    public Signal Signal { get; set; } = null!;
}

[Table("signal_events")]
public class SignalEvent
{
    [Key]
    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
    public long Id { get; set; }

    public long SignalId { get; set; }

    [Required, MaxLength(50)]
    public string EventType { get; set; } = string.Empty;

    public DateTime EventTs { get; set; } = DateTime.UtcNow;

    [Column(TypeName = "jsonb")]
    public string? PayloadJson { get; set; }

    [ForeignKey(nameof(SignalId))]
    public Signal Signal { get; set; } = null!;
}
